lib/web/sign.go

```diff
@@ -0,0 +1,205 @@
+		Writer:               virtualFS,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	archiveName := fmt.Sprintf("teleport_mTLS_%s.tar.gz", req.Hostname)
+
+	archiveBytes, err := archiveFromFiles(filesWritten, virtualFS)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Set file name
+	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment;filename="%v"`, archiveName))
+
+	// ServeContent sets the correct headers: Content-Type, Content-Length and Accept-Ranges.
+	// It also handles the Range negotiation
+	http.ServeContent(w, r, archiveName, time.Now(), bytes.NewReader(archiveBytes.Bytes()))
+
+	return nil, nil
+}
+
+// archiveFromFiles builds a Tar Gzip archive in memory, reading the files from the virtual FS
+func archiveFromFiles(files []string, virtualFS identityfile.InMemoryConfigWriter) (*bytes.Buffer, error) {
+	archiveBytes := &bytes.Buffer{}
+
+	gzipWriter := gzip.NewWriter(archiveBytes)
+	defer gzipWriter.Close()
+
+	tarWriter := tar.NewWriter(gzipWriter)
+	defer tarWriter.Close()
+
+	for _, filename := range files {
+		bs, err := virtualFS.Read(filename)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if err := tarWriter.WriteHeader(&tar.Header{
+			Name: filename,
+			Size: int64(len(bs)),
+
+			// https://www.postgresql.org/docs/current/libpq-ssl.html
+			// On Unix systems, the permissions on the private key file must disallow any access to world or group;
+			//  achieve this by a command such as chmod 0600 ~/.postgresql/postgresql.key.
+			// Alternatively, the file can be owned by root and have group read access (that is, 0640 permissions).
+			Mode: 0600,
+		}); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if _, err := tarWriter.Write(bs); err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	return archiveBytes, nil
+}
+
+type signCertKeyPairReq struct {
+	Hostname string
+	Format   identityfile.Format
+	TTL      time.Duration
+}
+
+// TODO(marco): only format db is supported
+var supportedFormats = []identityfile.Format{
+	identityfile.FormatDatabase,
+}
+
+func parseSignCertKeyPair(r *http.Request) (*signCertKeyPairReq, error) {
+	reqRaw := struct {
+		Hostname string `json:"hostname,omitempty"`
+		Format   string `json:"format,omitempty"`
+		TTL      string `json:"ttl,omitempty"`
+	}{}
+	if err := httplib.ReadJSON(r, &reqRaw); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ret := &signCertKeyPairReq{}
+
+	ret.Hostname = reqRaw.Hostname
+
+	if reqRaw.Format == "" {
+		return nil, trace.BadParameter("missing format")
+	}
+	ret.Format = identityfile.Format(reqRaw.Format)
+	if !slices.Contains(supportedFormats, ret.Format) {
+		return nil, trace.BadParameter("invalid format")
+	}
+
+	if reqRaw.TTL == "" {
+		reqRaw.TTL = apidefaults.CertDuration.String()
+	}
+	ttl, err := time.ParseDuration(reqRaw.TTL)
+	if err != nil {
+		return nil, trace.BadParameter("invalid ttl (please use https://pkg.go.dev/time#ParseDuration notation)")
+	}
+	ret.TTL = ttl
```

Comment:
We use `CheckAndSetDefaults` pattern so let's define a `CheckAndSetDefaults` method on the signing request that validates it and call it here. Grep the code for examples.